tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -494,23 +470,91 @@ const DeviceCommand EnqueueWriteBufferCommand::assemble_device_command(uint32_t
     command.set_consumer_cb_num_pages(consumer_cb_num_pages);
     command.set_num_pages(num_pages);
 
-    if (is_sharded(this->buffer.buffer_layout())) {
-        uint32_t num_cores = this->buffer.num_cores();
-        auto core_bank_indices = this->buffer.core_bank_indices();
-        command.set_sharded_buffer_num_cores(num_cores);
-    }
+
     command.set_data_size(padded_page_size * num_pages);
+    return command;
 
-    TT_ASSERT(padded_page_size <= consumer_cb_size, "Page is too large to fit in consumer buffer");
+}
 
-    return command;
+const DeviceCommand EnqueueWriteBufferCommand::assemble_device_command_sharded(uint32_t src_address) {
+    DeviceCommand command;
+    uint32_t num_pages = this->pages_to_write;
+    uint32_t buffer_address = this->buffer.address();
+    uint32_t padded_page_size = this->buffer.page_size();
+    if (this->buffer.page_size() != this->buffer.size()) { // should buffer.size() be num_pages * page_size
+        padded_page_size = align(this->buffer.page_size(), 32);
+    }
+    uint32_t src_page_index = 0;
+
+    uint32_t num_cores = this->buffer.num_cores();
+    uint32_t shard_size = this->buffer.shard_spec().size();
+    //TODO: for now all shards are same size of pages
+    vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
+    vector<uint32_t> core_id_x;
+    core_id_x.reserve(num_cores);
+    vector<uint32_t> core_id_y;
+    core_id_y.reserve(num_cores);
+    auto all_cores = this->buffer.all_cores();
+    for (const auto & core: all_cores) {
+        CoreCoord physical_core = device->worker_core_from_logical_core(core);
+        core_id_x.push_back(physical_core.x);
+        core_id_y.push_back(physical_core.y);
+    }
+    command.add_buffer_transfer_instruction_sharded(
+        src_address,
+        buffer_address,
+        num_pages,
+        padded_page_size,
+        (uint32_t) BufferType::SYSTEM_MEMORY,
+        (uint32_t) this->buffer.buffer_type(),
+        src_page_index,
+        this->dst_page_index,
+        num_pages_in_shards,
+        core_id_x,
+        core_id_y
+    );
+
+    command.set_sharded_buffer_num_cores(num_cores);
+
+    return write_assemble_device_command_postamble(command, padded_page_size, num_pages);
+}
+
+
+const DeviceCommand EnqueueWriteBufferCommand::assemble_device_command(uint32_t src_address) {
+    DeviceCommand command;
+    uint32_t num_pages = this->pages_to_write;
+    uint32_t buffer_address = this->buffer.address();
+    uint32_t padded_page_size = this->buffer.page_size();
+    if (this->buffer.page_size() != this->buffer.size()) { // should buffer.size() be num_pages * page_size
+        padded_page_size = align(this->buffer.page_size(), 32);
+    }
+    uint32_t src_page_index = 0;
```

Comment:
you need to rebase on main. Almeet added functionality for writing/reading offsets.

Reply:
@abhullar-tt could you point me to this commit. I tried rebasing and nothing changed on my end.

Reply:
nvm about this